feat(nhost-python): Python SDK (#4633)
nhost/nhost: docs/src/content/docs/getting-started/tutorials/python/3-graphql-operations.mdx

  is_pinned   boolean NOT NULL DEFAULT false,
  is_archived boolean NOT NULL DEFAULT false,
  created_at  timestamptz NOT NULL DEFAULT now(),
  updated_at  timestamptz NOT NULL DEFAULT now()
);

CREATE TABLE public.tags (
  id      uuid PRIMARY KEY DEFAULT gen_random_uuid(),
  user_id uuid NOT NULL REFERENCES auth.users (id) ON DELETE CASCADE,
  name    text NOT NULL,
  color   text NOT NULL DEFAULT '#808080',
  UNIQUE (user_id, name)
);

CREATE TABLE public.note_tags (
  note_id uuid NOT NULL REFERENCES public.notes (id) ON DELETE CASCADE,
  tag_id  uuid NOT NULL REFERENCES public.tags (id) ON DELETE CASCADE,
  PRIMARY KEY (note_id, tag_id)
);

CREATE TRIGGER set_notes_updated_at BEFORE UPDATE ON public.notes
  FOR EACH ROW EXECUTE FUNCTION public.set_updated_at();
```

Make sure **Track this** is enabled for each table so it's exposed over GraphQL.
Then open each table's **Relationships** tab and track the foreign-key
relationships Nhost suggests — including the two on the `note_tags` join table:
the object relationship to `notes` (**name it `note`**) and the one to `tags`
(**name it `tag`**). The `note` name matters — the `note_tags` permission below

> references it by that exact name, so a mismatch produces `"note" does not exist`.

Reviewer: The tutorial's queries select the `notes → note_tags` array relationship as **`noteTags`** (below), but this paragraph only pins the `note`/`tag` names. Nhost suggests that array relationship as `note_tags`, so a reader who tracks the suggestions verbatim then hits `field 'noteTags' not found in type: 'notes'` on `note ls`. Either pin the `noteTags` name here (suggested) or switch the code to `note_tags`.

```suggestion
relationships Nhost suggests — including the two on the `note_tags` join table:
the object relationship to `notes` (**name it `note`**) and the one to `tags`
(**name it `tag`**). Also rename the `notes` table's array relationship to
`note_tags` to **`noteTags`** — the note-list query below selects it by that
camelCase name. These names matter: the `note_tags` permission references
`note` and the note-list query references `noteTags`, so a mismatch produces
`"note" does not exist` or `field 'noteTags' not found`.
```

filter you configured means the database only ever returns the caller's rows.

## Update, tag, and organise

Updates go through `update_notes_by_pk`. One helper covers edit, pin, archive,
and move; an empty primary-key result means the row doesn't exist *or* isn't
yours:

```python title="main.py"
async def update_note(nhost, note_id: str, set_: dict) -> None:
    data = await gql(
        nhost,
        """
        mutation UpdateNote($id: uuid!, $set: notes_set_input!) {
          update_notes_by_pk(pk_columns: {id: $id}, _set: $set) { id }
        }""",
        {"id": note_id, "set": set_},
    )
    if not data.get("update_notes_by_pk"):
        raise SystemExit("note not found or not permitted")
    print("updated", note_id)
```

Tags use an **upsert**: `insert_tags_one` with `on_conflict` reuses an existing
tag (unique per `user_id, name`) instead of failing, then `insert_note_tags_one`
links it to the note:

```python title="main.py"
async def upsert_tag(nhost, name: str) -> str:
    data = await gql(
        nhost,
        """
        mutation UpsertTag($name: String!) {
          insert_tags_one(
            object: {name: $name}

>             on_conflict: {constraint: tags_user_id_name_key, update_columns: []}

Reviewer: `update_columns: []` compiles to `ON CONFLICT DO NOTHING`, so when a tag already exists Hasura returns `null` for `insert_tags_one` and `data["insert_tags_one"]["id"]` (line 217) raises `TypeError: 'NoneType' object is not subscriptable`. This fires whenever a user reuses a tag. `update_columns: [name]` makes it a no-op `DO UPDATE` (sets `name` to the same conflicting value) so the existing row is returned.

_Not `[color]` — `excluded.color` would reset a customized tag color to the default._

```suggestion
            on_conflict: {constraint: tags_user_id_name_key, update_columns: [name]}
```

    data = await gql(
        nhost,
        """
        query Notes($where: notes_bool_exp!) {
          notes(where: $where, order_by: [{is_pinned: desc}, {updated_at: desc}]) {
            id title is_pinned notebook { name } noteTags { tag { name } }
          }
        }""",
        {"where": where},
    )
    notes = data.get("notes", [])
    if not notes:
        print("(no notes)")
        return
    for n in notes:
        pin = "*" if n["is_pinned"] else " "
        print(f"{pin} {n['id']}  {n['title']}")

async def update_note(nhost, note_id: str, set_: dict) -> None:
    data = await gql(
        nhost,
        """
        mutation UpdateNote($id: uuid!, $set: notes_set_input!) {
          update_notes_by_pk(pk_columns: {id: $id}, _set: $set) { id }
        }""",
        {"id": note_id, "set": set_},
    )
    if not data.get("update_notes_by_pk"):
        raise SystemExit("note not found or not permitted")
    print("updated", note_id)

async def upsert_tag(nhost, name: str) -> str:
    data = await gql(
        nhost,
        """
        mutation UpsertTag($name: String!) {
          insert_tags_one(
            object: {name: $name}

>             on_conflict: {constraint: tags_user_id_name_key, update_columns: []}

Reviewer: Same `update_columns: []` upsert bug in the full-`main.py` listing.

```suggestion
            on_conflict: {constraint: tags_user_id_name_key, update_columns: [name]}
```
